src/shared/Core/Trace2.cs

```diff
@@ -0,0 +1,93 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+using Newtonsoft.Json.Serialization;
+
+namespace GitCredentialManager;
+
+// The different event types tracked in the TRACE2 tracing
+// system.
+public enum Event {}
+
+/// <summary>
+/// Represents the application's TRACE2 tracing system, which
+/// writes to a collector service that stores the trace information.
+/// </summary>
+public interface ITrace2 : IDisposable
+{
+    /// <summary>
+    /// Add a listener to the trace writer.
+    /// </summary>
+    /// <param name="listener">The listener to add.</param>
+    void AddListener(CollectorWriter listener);
+}
+
+public class Trace2: DisposableObject, ITrace2
+{
+    private const string ParentSidVariable = "GIT_TRACE2_PARENT_SID";
+    private const string CurrentSidVariable = "GCM_TRACE2_PARENT_SID";
+    private readonly ICommandContext _context;
+    private bool _isDisposed;
+    private CollectorWriter _writer;
+
+    public void AddListener(CollectorWriter listener)
+    {
+        ThrowIfDisposed();
+        
+        // Try not to add the same listener more than once
+        if (_writer == listener)
+            return;
+
+        _writer = listener;
+        _writer.Start();
+    }
+
+    public new void Dispose()
+    {
+        if (_isDisposed)
+        {
+            // This instance has already been disposed
+            return;
+        }
+        
+        _writer.Dispose();
+        _isDisposed = true;
+    }
```

Comment:
If we're inheriting from `DisposableObject` we shouldn't be hiding the base `Dispose` method (using `new`).

`DisposableObject` implements the `IDisposable` pattern and makes it easier to implement by offering two extension points:

```csharp
/// <summary>
/// Called when unmanaged resources should be released and memory freed.
/// </summary>
protected virtual void ReleaseUnmanagedResources() { }

/// <summary>
/// Called when managed resources should be released.
/// </summary>
protected virtual void ReleaseManagedResources() { }
```

Here you'd just want to `override` `ReleaseManagedResources` to dispose of `_writer`.


```csharp
protected override void ReleaseManagedResources()
{
    _writer.Dispose();
    base.ReleaseManagedResources();
}
```